lib/net/nforceif/include/lwipopts.h

```diff
@@ -420,6 +420,59 @@
  */
 #define LWIP_STATS                      1
 
+/*
+   ---------------------------------------
+   ---------- Hook options ---------------
+   ---------------------------------------
+*/
+
+/**
+ * LWIP_HOOK_DHCP_APPEND_OPTIONS(netif, dhcp, state, msg, msg_type, options_len_ptr):
+ * Called from various dhcp functions when sending a DHCP message.
+ * This hook is called just before the DHCP message trailer is added, so the
+ * options are at the end of a DHCP message.
+ * Signature:\code{.c}
+ *   void my_hook(struct netif *netif, struct dhcp *dhcp, u8_t state, struct dhcp_msg *msg,
+ *                u8_t msg_type, u16_t *options_len_ptr);
+ * \endcode
+ * Arguments:
+ * - netif: struct netif that the packet will be sent through
+ * - dhcp: struct dhcp on that netif
+ * - state: current dhcp state (dhcp_state_enum_t as an u8_t)
+ * - msg: struct dhcp_msg that will be sent
+ * - msg_type: dhcp message type to be sent (u8_t)
+ * - options_len_ptr: pointer to the current length of options in the dhcp_msg "msg"
+ *                    (must be increased when options are added!)
+ *
+ * Options need to appended like this:
+ *   LWIP_ASSERT("dhcp option overflow", *options_len_ptr + option_len + 2 <= DHCP_OPTIONS_LEN);
+ *   msg->options[(*options_len_ptr)++] = &lt;option_number&gt;;
+ *   msg->options[(*options_len_ptr)++] = &lt;option_len&gt;;
+ *   msg->options[(*options_len_ptr)++] = &lt;option_bytes&gt;;
+ *   [...]
+ */
+/*
+   Add DHCP options to imitate XDK behavior, making DHCP servers treat nxdk the same,
+   i.e. recognizing requests as coming from the same device and assigning the same
+   IP address.
+*/
+#define LWIP_HOOK_DHCP_APPEND_OPTIONS(netif, dhcp, state, msg, msg_type, options_len_ptr) \
+    *options_len_ptr = dhcp_option(*options_len_ptr, msg->options, DHCP_OPTION_CLIENT_ID, 7); \
```

Comment:
Generally I'd like to avoid macro code blocks like this. Similar to how there's a `dhcp_option_hostname` function, I think this code could naturally be broken up into two functions, e.g. `dhcp_option_client_id`, `dhcp_option_vendor_class_id`, the vendor string converted to a loop over a `char *dhcp_option_vendor_class_id_str`, and this hook macro would simply call a third, wrapper function to append all the args....or something like that. Upstream LWIP might even want this feature.

Reply:
Usually I'd prefer to put this into functions, too.
However, `lwipopts.h` is included in lots of different places, of which not all include the necessary headers to have all types available, so I can't just move the code into a static function here, unless I make sure to include all those headers, which I'd like to avoid as `lwipopts.h` isn't meant to do that.
Even if I put the function into a separate C-file (which I'd like to avoid, too), I'd still have to pull in those headers for the declaration.

Reply:
> However, lwipopts.h is included in lots of different places, of which not all include the necessary headers to have all types available, so I can't just move the code into a static function here, unless I make sure to include all those headers, which I'd like to avoid as lwipopts.h isn't meant to do that.

I agree it's not ideal, but this macro definition is specifically designed to add code, so pulling in the headers in the appropriate place would be acceptable in my opinion. Frankly, it would probably just be better to patch LWIP to add this natively, instead of adding as a hook like this. For now, let's just keep it simple and use the approach you've submitted and we can refactor it later.